File: VanillaMCTS/mcts.py
from collections import namedtuple
import connect4 as c4
import copy
import math
import random

State = namedtuple('State', ['board', 'turn'])

class Node:
    def __init__(self, state, parent=None, parent_action=None):

        self.state = state
        self.parent = parent
        self.parent_action = parent_action

        self.children = {}
        

        self.untried_moves = c4.get_legal_moves(state.board)
        self.Na = dict.fromkeys(self.untried_moves, 0)
        self.Wa = dict.fromkeys(self.untried_moves, 0.0)
        self.Qa = dict.fromkeys(self.untried_moves, 0.0)
        self.N = 0


tokens = ['X', 'O']

def UCT(node, action, c):
    if node.Na[action] == 0:
        return float("inf")
    return node.Qa[action] + c*math.sqrt(math.log(node.N) / node.Na[action])

def select(node, c=math.sqrt(2)):
    if node.untried_moves:
        return node
    if not node.children:
        return node
    return select(node.children[max(node.children, key=lambda action: UCT(node, action, c))])

def expand(node):
    action = random.choice(tuple(node.untried_moves))
    node.untried_moves.remove(action)
    new_state = copy.deepcopy(node.state)
    c4.drop_token(new_state.board, action, tokens[new_state.turn])
    new_state = new_state._replace(turn=1 - new_state.turn)
    new_node = Node(new_state, parent=node, parent_action=action)
    node.children[action] = new_node
    return new_node

def simulate(state):
    board = copy.deepcopy(state.board)
    turn = state.turn
    while True:
        if c4.check_win(board, tokens[turn]):
            return -1 if state.turn == turn else 1
        if c4.is_tie(board):
            return 0
        turn = 1 - turn
        action = random.choice(tuple(c4.get_legal_moves(board)))
        c4.drop_token(board, action, tokens[turn])

def backprop(node, result):
    node.N += 1
    if node.parent is not None:
        parent = node.parent
        a = node.parent_action
        parent.Na[a] += 1
        parent.Wa[a] += result
        parent.Qa[a] = parent.Wa[a] / parent.Na[a]
        backprop(parent, -result)


def MCTS(root, simulations=1000):
    for _ in range(simulations):
        node = select(root)
        leaf = expand(node) if node.untried_moves else node
        result = simulate(leaf.state)
        backprop(leaf, result)
    return max(root.children.items(), key=lambda x: x[1].N)[0]

File: engine/mcts_cpp/src/mcts.cpp
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <vector>
#include <random>
#include <cmath>
#include "mcts.h"

namespace py = pybind11;

struct Node 
{
    py::object state;
    std::vector<py::object> moves;
    std::vector<int> Na;
    std::vector<double> Wa;
    std::vector<double> Qa;
    std::vector<Node*> children;
    std::vector<int> untried;
    Node* parent;
    int parent_action_idx;
    int N;

    Node(py::object st, const py::list& move_list, Node* p=nullptr, int idx=-1): state(st), parent(p), parent_action_idx(idx), N(0) 
    {
        for (auto m : move_list) moves.push_back(py::reinterpret_borrow<py::object>(m));
        size_t n = moves.size();
        Na.assign(n, 0);
        Wa.assign(n, 0.0);
        Qa.assign(n, 0.0);
        children.assign(n, nullptr);
        for (size_t i=0;i<n;++i) untried.push_back(i);
    }
    ~Node() 
    {
        for (Node* child : children) delete child;
    }
};

static double UCT(const Node* node, int a, double c) 
{
    if (node->Na[a] == 0) return std::numeric_limits<double>::infinity();
    return node->Qa[a] + c * std::sqrt(std::log((double)node->N) / node->Na[a]);
}

static Node* select(Node* node, double c) 
{
    while (node) 
    {
        if (!node->untried.empty()) return node;
        int best = -1; double best_val=-1e100;
        for (size_t i=0;i<node->children.size();++i) 
        {
            if (!node->children[i]) continue;
            double v = UCT(node, i, c);
            if (v > best_val) { best_val = v; best = (int)i; }
        }
        if (best == -1) return node;
        node = node->children[best];
    }
    return node;
}

static Node* expand(Node* node, py::object backend, py::object policy) 
{
    py::list untried_moves;
    for (int idx : node->untried) untried_moves.append(node->moves[idx]);
    py::object action = policy(untried_moves);
    int local_idx = untried_moves.attr("index")(action).cast<int>();
    int move_idx = node->untried[local_idx];
    node->untried.erase(node->untried.begin()+local_idx);
    py::object new_state = backend.attr("play_move")(node->state, action);
    py::list new_moves = backend.attr("get_legal_moves")(new_state);
    Node* child = new Node(new_state, new_moves, node, move_idx);
    node->children[move_idx] = child;
    return child;
}

static void backprop(Node* node, double result) 
{
    while (node) 
    {
        node->N += 1;
        if (node->parent) 
        {
            int a = node->parent_action_idx;
            Node* parent = node->parent;
            parent->Na[a] += 1;
            parent->Wa[a] -= result;
            parent->Qa[a] = parent->Wa[a] / parent->Na[a];
            node = parent;
            result = -result;
        } 
        else 
        {
            break;
        }
    }
}

py::object get_move(py::object state, py::object value, py::object policy, py::object backend, int simulations, double c, int batch_size) 
{
    py::list moves = backend.attr("get_legal_moves")(state);
    Node* root = new Node(state, moves);
    std::vector<Node*> pending_nodes;
    std::vector<py::object> pending_states;
    auto flush = [&]() 
    {
        if (pending_nodes.empty()) return;
        py::gil_scoped_acquire gil;
        py::list states;
        for (auto& s : pending_states) states.append(s);
        py::object vals_obj = value.attr("batch")(states, py::arg("backend")=backend);
        auto vals = vals_obj.cast<py::list>();
        for (size_t i=0;i<pending_nodes.size();++i) 
        {
            double v = vals[i].cast<double>();
            backprop(pending_nodes[i], v);
        }
        pending_nodes.clear();
        pending_states.clear();
    };

    for (int i=0;i<simulations;i++) 
    {
        Node* node = select(root, c);
        Node* leaf;
        if (!node->untried.empty()) 
        {
            py::gil_scoped_acquire gil;
            leaf = expand(node, backend, policy);
        } 
        else 
        {
            leaf = node;
        }
        pending_nodes.push_back(leaf);
        pending_states.push_back(leaf->state);
        if ((int)pending_nodes.size() >= batch_size) 
        {
            flush();
        }
    }
    flush();
    int best_idx=-1; int best_N=-1;
    for (size_t i=0;i<root->children.size();++i) 
    {
        Node* child = root->children[i];
        if (child && child->N > best_N) { best_N = child->N; best_idx = (int)i; }
    }
    py::object best_move = root->moves[best_idx];
    delete root;
    return best_move;
}

